### DISCUSSION TITLE
Several issues introduced in version 20230306 on audio with silences (repeated text, segment "id" not unique/increasing)

### DISCUSSION THREAD

**Discussion by Jeronymous** [openai/whisper#1058] [Q&A] [answered]
This issue looks more general than https://github.com/openai/whisper/discussions/1046 and it might be related to https://github.com/openai/whisper/discussions/730 (however this was after v20230306).

---

Take this audio:  [bonjour_vous_allez_bien.mp3](https://github.com/openai/whisper/files/10919955/bonjour_vous_allez_bien.zip)
where "Bonjour, est-ce que vous allez bien?" (French) is said twice, with a long delay in between.

For this command:
```
whisper --model tiny --language fr \
   --beam_size None --temperature_increment_on_fallback None --best_of None \
    bonjour_vous_allez_bien.mp3
```
below are the differences between outputs of previous version 20230124 (left), new version 20230306 (middle) and 20230306 with `--word_timestamps True` (right). We can see two issues:
* 20230306 is producing a lot of hallucinated repetitions and this is much worse with --word_timestamps True
* the "id" of each segment is not necessarily unique in 20230306. It can be several times the same in a row (it gets stuck at "0" in the example below)
![image](https://user-images.githubusercontent.com/22522728/223698216-7f5efeae-f8db-4552-9406-5abdd9562db1.png)

---

This became obvious when I used greedy decoding, but a similar thing can be observed with less options (with beam search).
```
whisper --model tiny --language fr bonjour_vous_allez_bien.mp3
```
On this command, the differences between outputs of previous version 20230124 (left), new version 20230306 (middle) and 20230306 with `--word_timestamps True` (right) are:

![image](https://user-images.githubusercontent.com/22522728/223700680-e574c858-e115-4ad1-8601-911676f59e4b.png)

---

I also noticed that the following command is particularly long to run:
```
whisper --model tiny --language fr --word_timestamps True /home/jlouradour/src/whisper-timestamped/tests/data/bonjour_vous_allez_bien.mp3
```
It takes 1 minute with 4 CPU, whereas it takes 6 sec without `--word_timestamps True`, or 16 sec if adding options `--beam_size None --temperature_increment_on_fallback None --best_of None`

**Comment by Jeronymous** [comment]
I've just tested 20230307 and the two problems are still here:
- not unique "id" (not necessarily increasing)
- the "text" field of the transcription include hallucinated repetition, bu actually that those are not present in the segments (so if we join all the "text" of the segments, we get a result that is different from the main "text")

**Reply by glangford** [reply]
This looks like the repetition is unique to the .json output, the others (.txt, .srt, etc) look correct. Do you agree?

**Reply by glangford** [reply]
I can reproduce this problem only with the tiny model, and only for the .json output. Small and medium and other outputs seem to be ok. `word_timestamps` does not have to be `True`.
`whisper --model tiny --language fr bonjour_vous_allez_bien.mp3`

If I don't supply `--language`, the tiny model detects the audio as Russian.

**Reply by SalimovAlbert** [reply]
Had the same issue with JSON, used  `"".join(map(lambda x: x['text'], result['segments']))`  to get text output

**Reply by Jeronymous** [reply]
> This looks like the repetition is unique to the .json output, the others (.txt, .srt, etc) look correct. Do you agree?

Yes, correct.
It only affects json output (or the output of the transcribe function when using python and not the CLI)

**Comment by jongwook** [comment]
Hi! Thanks for reporting this. The non-unique `"id"` is my total oversight and should be an easy fix; wasn't paying attention to it much.

About the discrepancy between ["text"] and segment["text"], I think it has something to do with these lines:

https://github.com/openai/whisper/blob/aac47c98349b98cec5ca7b1be53960fb59f4436b/whisper/transcribe.py#L345-L356

EDIT: the culprit was actually this line:

https://github.com/openai/whisper/blob/aac47c98349b98cec5ca7b1be53960fb59f4436b/whisper/transcribe.py#L290

will push a fix soon.

**Comment by jongwook** [comment, marked as answer]
This was an issue where the new `transcribe()` was mishandling the `all_tokens` variable which affected the prompts to be more prone to repetitions and also caused the discrepancy between the top-level `"text"` field and the segment-level `"text"` fields in the JSON response.

I have a fix merged in #1060, so hopefully this resolves your issue! Please let me know if it continues.

**Reply by Jeronymous** [reply]
Great. Thank you @jongwook

**Comment by timothyaveni** [comment]
I think the `condition_on_previous_text` flag was recently broken, and my hunch is that the cause is some changes in #1060; in particular, I think `all_tokens` is off by one window at the point where the prompt is reset, causing the prior window's tokens to be added to the prompt anyway. @jongwook would you mind having a look? Thanks!

**Reply by guillaumekln** [reply]
A PR was opened to fix this issue: #1224.

**Reply by timothyaveni** [reply]
oop, didn't think to look there. thanks for the pointer!
